### PR TITLE
Fix: broken contract in StoreAPI ProductCollectionData endpoint

### LINKED ISSUE #42157: StoreAPI: Query parameters via Collection Data endpoint are no longer working
Hey there!

when updating to the latest WooCommerce version my StoreAPI request data doesn't seem to return the data it used to return in WC 7.6.

I'm using the [collection-data](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/product-collection-data.md) API to get a pre-filtered result set based on a given categories in a custom ACF block which shows products on any page it is placed on. So the products are not part of the main query.

It doesn't seem to work properly anymore since the [this](https://github.com/woocommerce/woocommerce-blocks/pull/8599) PR seem to dropped support for `ProductQuery()`  which will no longer take the category query parameter into account.

Here's a screenshot of my `WP_REST_Request` instance but the request returns all attributes instead of the ones filtered for the given category terms
image

I noticed the Collection API docs still state support for all the parameters
> In addition to the above attributes, all product list attributes are supported. This allows you to get data for a certain subset of products. See [the products API list products section](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/products.md#list-products) for the full list.

Is this a bug or am I missing something here? Any help would be appreciated here Thank you! 🙌

### PR DESCRIPTION
**Pull request by dinhtungdu**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

 I revisited https://github.com/woocommerce/woocommerce-blocks/issues/7607 and found that:

- The Attribute Filter block and original `get_attribute_counts` method worked fine. The block passed the correct data data to the endpoint. The `ProductCollectionData` processed the passed param correctly.
- The Attribute Filters block also worked as expected originally. It has the count mismatch issue when used with the price filter.
- What makes the count mismatch in https://github.com/woocommerce/woocommerce-blocks/issues/7607 is an issue in the price clause, which still exists in `trunk`:

https://github.com/woocommerce/woocommerce/blob/b05082ae874688db69a986cc03d2691b881a7163/plugins/woocommerce/src/StoreApi/Utilities/ProductQuery.php#L404-L422

What is wrong with this logic? In the meta lookup table, each product has a min and a max price. If the variations of a product have different prices, then the min and max prices will be set to the lowest and highest prices among variations. Otherwise, min and max prices will be identical.

If a `min_price` filter is set, we need to look for products that have a maximum price greater or equal to that filter. Similar to `max_price`, filtered products need to have a minimum price smaller or equal to the filter price.

We're doing the reverse in `ProductQuery::add_price_filter_clauses`, fixing the price clause, StoreAPI returns the correct counts for filter blocks (for all products/product catalog page).

Closes #42157.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Verify this fix the broken contract:

1. Make a request to the `product-collection-data` endpoint of the StoreAPI with the `category` params containing the IDs of some categories.
2. See the returned products belonging to the passed categories only.
3. Try other requests with other attributes like `on_sale`, see only on-sale product IDs returned.
 
#### Ensure we don't introduce any regression:

4. Add filter blocks (current React-based ones) using the `Product Filters` pattern to the Product Catalog template.
5. Ensure the Product Catalog template contains a Classic Template or Products/Product Collection block that inherits query from the template.
6. Testing the Product Catalog page against test cases in  https://github.com/woocommerce/woocommerce-blocks/issues/7607 and https://github.com/woocommerce/woocommerce-blocks/pull/8599. Ensure all test case passes.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by nefeline — COMMENTED** [review]
Thanks, Tung!

There are a handful of problems with the counts, listing a few:

`?min_price=19&max_price=31&filter_size=large%2Cmedium&query_type_size=or`

Screenshot 2024-03-02 at 07 07 39

`?filter_stock_status=instock&max_price=18&rating_filter=4`

Screenshot 2024-03-02 at 07 25 00

`?rating_filter=5&max_price=15`

Screenshot 2024-03-02 at 07 19 09

Tagging @mikejolley @roykho @Konamiman @vedanshujain here as given the extent of the changes proposed, I believe it's important to have more eyes from engineers who have worked on this area of the codebase to sign off this update, if they have the availability, of course :)

@dinhtungdu considering that I've already flagged my concerns with this approach on https://github.com/woocommerce/woocommerce/discussions/44936 and have just provided additional feedback on this PR as well, I'm waiving my review here moving forward as at this point I believe it's important to gather the opinions of other folks so we can ensure that whatever path we choose for this implementation, it is the best for store owners.

**Comment by dinhtungdu** [comment]
@nefeline Thanks for testing!

I can reproduce the issues you mentioned above. At first, I also think it's the count data issue. But looking at the network tab, I see something interesting. Take your first case as the example: `?min_price=19&max_price=31&filter_size=large%2Cmedium&query_type_size=or`.

There are three requests to the `product-collection-data` endpoint, here are their payload and response (trimmed to `attribute_counts`):

The first one:

image
image

The second one:

image
image

The third one:
image
image

The third one is the complete request we want to use, and the response is correct as well. From the results, I think:

- StoreAPI/`product-collection-data` is working as expected and returns the correct count.
- There is an issue with the way we consolidate and fetch the count data in our filter blocks. In this case, the attribute filter block displays the result from the initial requests, not the last one which is the complete one with the correct count. 

https://github.com/woocommerce/woocommerce/pull/45247/commits/e8595ba3f9d37baa6c14af30271882af457e82ba is my attempt to fix the issues. I decided to always send the `attributes` param to the endpoint because we already have a logic handle that param in the `ProductCollectionData` class, so we don't have to do the same on the client side. Passing the raw `queryState` to the endpoint, I can see the filter blocks display the correct count now. Filter blocks also still work as expected with the `All Products` block.

https://github.com/woocommerce/woocommerce/pull/45247/commits/e8595ba3f9d37baa6c14af30271882af457e82ba is more a workaround than a fix to me. I still can't control the last request to fetch and data to be set for the last render. But as we're deprecating the current filter blocks for new ones, I think a workaround is sufficient enough. Updating `useCollection` and/or `useCollectionData` is high risk because we're using them in multiple places, which isn't worth the effort in this case, for to-be-deprecated blocks, IMO.

> considering that I've already flagged my concerns with this approach on https://github.com/woocommerce/woocommerce/discussions/44936 and have just provided additional feedback on this PR as well, I'm waiving my review here moving forward as at this point

I understand your concerns, but I believe this can fix the broken contract without reintroducing the regression. Would you mind testing this PR again with https://github.com/woocommerce/woocommerce/pull/45247/commits/e8595ba3f9d37baa6c14af30271882af457e82ba to see if it fixes the issues mentioned above?

> I believe it's important to gather the opinions of other folks so we can ensure that whatever path we choose for this implementation, it is the best for store owners.

I totally agree with this!

**Comment by nefeline** [comment]
Thanks a lot for the hard work you've put into this. As mentioned previously, I will step back as it might be best to get input from others who might be more in sync with your proposed direction at this point.

**Review by roykho — DISMISSED** [review]
Thanks for fixing the regression.

> But as we're deprecating the current filter blocks for new ones, I think a workaround is sufficient enough

I agree with this.

I've tested this with several combinations and they all seem to work as expected. Moreover testing with the limited product categories now indeed does return only the products/counts within those categories.

I left some inline comments.

File: plugins/woocommerce/src/StoreApi/Routes/V1/ProductCollectionData.php [hunk 1]
```diff
@@ -132,31 +169,31 @@ public function get_collection_params() {
 		$params = ( new Products( $this->schema_controller, $this->schema ) )->get_collection_params();
 
 		$params['calculate_price_range'] = [
-			'description' => __( 'If true, calculates the minimum and maximum product prices for the collection.', 'woocommerce' ),
+			'description' => __( 'If true, calculates the minimum and maximum product prices for the collection.', 'woo-gutenberg-products-block' ),
```
Comment (roykho) [inline plugins/woocommerce/src/StoreApi/Routes/V1/ProductCollectionData.php]: I believe the text domain should be `woocommerce` right (since the merge)?

Comment (thealexandrelara) [reply plugins/woocommerce/src/StoreApi/Routes/V1/ProductCollectionData.php]: Nice catch! I just fixed this in https://github.com/woocommerce/woocommerce/pull/45247/commits/6e473aded759a4e5949328c25d6d5f75913efa39

**Comment by roykho** [comment]
I am seeing tests failing. More specifically the attributes counts. Perhaps we need to update the tests?

**Comment by roykho** [comment]
@thealexandrelara - I have fixed the tests if you could give it a review and merge.

**Review by thealexandrelara — APPROVED** [review]
Awesome! Thank you for fixing the tests @roykho. I re-tested and everything is working as expected. I'm approving and merging it!

**Comment by sunyatasattva** [comment]
Yuhuu! Thank you so much folks for getting this over the finishing line!

**Pull request merged**